rekall-core/rekall/plugins/linux/heap_analysis.py

```diff
@@ -0,0 +1,4388 @@
+                )
+
+        return None
+
+    def _get_number_of_cores(self):
+        """Returns the number of cpu cores for the current memory image."""
+
+        return len(list(cpuinfo.CpuInfo(session=self.session).online_cpus()))
+
+
+    def _get_max_number_of_arenas(self):
+        """Returns the maximum number of supported arenas. This value depends on the number
+        of cpu cores."""
+
+        cores = self._get_number_of_cores()
+        return cores * (2 if self._size_sz == 4 else 8)
+
+
+    def _check_arenas(self, arena, deactivate_swap_check=False):
+        """Iterates the next field of the malloc_state struct and checks if we end up
+        at the same malloc_state after the maximum number of arenas for the current system.
+        Checks also for arena structs being part of swapped memory pages.
+        """
+
+        # This function is only reliable, if we have the offset to mp_
+        if not self.mp_:
+            # at least we test arena for being swapped
+            if not deactivate_swap_check:
+                self._check_and_report_arena_for_being_swapped(arena)
+
+            return None
+
+        # max arena value can be adjusted at runtime via mallopt func:
+        # see malloc/malloc.c line 4753 and http://man7.org/linux/man-pages/man3/mallopt.3.html
+        # or on startup via env vars (see also link)
+        # if not, this member is 0
+        arena_max = self.mp_.arena_max
+
+        if arena_max == 0:
+            # The maximum number of arenas is calculated with the macro 'NARENAS_FROM_NCORES'
+            # See malloc/arena.c
+            arena_max = self._get_max_number_of_arenas()
+            cores = self._get_number_of_cores()
+
+            # In the case of one core, there can be one more arena than
+            # the result from 'NARENAS_FROM_NCORES'
+            # See function 'arena_get2' in malloc/arena.c
+            if cores == 1:
+                arena_max += 1
+
+        if arena_max == 0:
+            self.session.logging.warn("The result for arena_max has been 0. This shouldn't be "
+                                      "the case and has to be looked into.")
+
+            if not deactivate_swap_check:
+                # as the following for loop will in this case not loop over any arena, we check
+                # the current arena at least for being swapped
+                self._check_and_report_arena_for_being_swapped(arena)
+
+        curr_arena = arena
+        for _ in range(arena_max):
+            swap_check_result = self._check_and_report_arena_for_being_swapped(curr_arena) if \
+                not deactivate_swap_check else None
+
+            if swap_check_result is not True:
+                curr_arena = curr_arena.next
+                if arena == curr_arena:
+                    return True
+
+            else:
+                break
+
+        return False
+
+
+
+    def __init__(self,
+                 glibc_profile=None,
```

Comment:
if a libc profile is provided as a file (via cmdline) then the plugin tries to use it for every process
if not, the glibc profile is loaded for each task separately, so it is able to use a different profile for each process